airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```diff
@@ -22,97 +24,95 @@
  * ContainerFactory is the companion interface to {@link TestDatabase} for providing it with
  * suitable testcontainer instances.
  */
-public interface ContainerFactory<C extends JdbcDatabaseContainer<?>> {
+public abstract class ContainerFactory<C extends JdbcDatabaseContainer<?>> {
 
-  /**
-   * Creates a new, unshared testcontainer instance. This usually wraps the default constructor for
-   * the testcontainer type.
-   */
-  C createNewContainer(DockerImageName imageName);
+  static private final Logger LOGGER = LoggerFactory.getLogger(ContainerFactory.class);
 
-  /**
-   * Returns the class object of the testcontainer.
-   */
-  Class<?> getContainerClass();
+  private record ContainerKey(Class<? extends ContainerFactory> clazz, DockerImageName imageName, List<String> methods) {};
 
-  /**
-   * Returns a shared instance of the testcontainer.
-   */
-  default C shared(String imageName, String... methods) {
-    final String mapKey = Stream.concat(
-        Stream.of(imageName, this.getClass().getCanonicalName()),
-        Stream.of(methods))
-        .collect(Collectors.joining("+"));
-    return Singleton.getOrCreate(mapKey, this);
-  }
+  private static class ContainerOrException {
 
-  /**
-   * This class is exclusively used by {@link #shared(String, String...)}. It wraps a specific shared
-   * testcontainer instance, which is created exactly once.
-   */
-  class Singleton<C extends JdbcDatabaseContainer<?>> {
+    private final Supplier<GenericContainer<?>> containerSupplier;
+    private volatile RuntimeException exception = null;
+    private volatile GenericContainer<?> _container = null;
```

Comment:
either remove this underscore or add one in the variable name above, for consistency